src/main/java/com/google/step/YOUR_PROJECT_NAME_HERE/servlets/SearchServlet.java

```diff
@@ -15,24 +15,99 @@
 package com.google.step.YOUR_PROJECT_NAME_HERE.servlets;
 
 import java.io.IOException;
-import java.util.Arrays;
+import java.io.UnsupportedEncodingException;
+import java.net.URI;
+import java.net.URLEncoder;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
 import javax.servlet.ServletException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import org.json.JSONObject;
 
 /** Servlet that handles searches. */
 @WebServlet("/search")
 public class SearchServlet extends HttpServlet {
 
+  private final HttpClient httpClient =
+      HttpClient.newBuilder().version(HttpClient.Version.HTTP_2).build();
+
+  private static String encodeValue(String value) {
+    try {
+      return URLEncoder.encode(value, StandardCharsets.UTF_8.toString());
+    } catch (UnsupportedEncodingException ex) {
+      throw new RuntimeException(ex.getCause());
+    }
+  }
+
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response)
       throws IOException, ServletException {
-    // Set some data on the request for the template to access.
-    request.setAttribute("previousQuery", request.getParameter("q"));
-    request.setAttribute("results", Arrays.asList("Aaron", "Iris", "Kayla"));
-    // Forward the request to the template (which is a servlet itself).
-    request.getRequestDispatcher("WEB-INF/templates/search.jsp").forward(request, response);
+
+    String query = encodeValue(request.getParameter("q"));
+    List<String> allLinks = new ArrayList<>();
+
+    String w3GoogleCSE =
+        "https://www.googleapis.com/customsearch/v1?key=AIzaSyCSKQy6t7HgEc1zI2thwgGncFVzyB8zZ00&cx=010448377421452380243:pat2rmwjvb8&q=";
+
+    String stackGoogleCSE =
+        "https://www.googleapis.com/customsearch/v1?key=AIzaSyCSKQy6t7HgEc1zI2thwgGncFVzyB8zZ00&cx=010448377421452380243:6b8rs1ze5oy&q=";
+
+    String geeksGoogleCSE =
+        "https://www.googleapis.com/customsearch/v1?key=AIzaSyCSKQy6t7HgEc1zI2thwgGncFVzyB8zZ00&cx=010448377421452380243:ieq7z84z2qq&q=";
+
+    try {
```

Comment:
I think a for loop here might be cleaner

Reply:
I think it'll be easier to keep it in this sequential non-loop structure for now, until scraping is fully implemented for each of the sites because it'll be easier for each of us to debug/work on the respective site scraping without the calls being in a loop. But we can revisit the idea of a loop to clean it up after scraping implementation is complete?

Reply:
Ah good thought, yeah in that case let's leave it this way

Reply:
Maybe add a `TODO: after implementing scraping, consider changing this to be a for-loop.` or something like that.